compiler/rustc_codegen_llvm/src/gotoc/rvalue.rs

```diff
@@ -94,82 +94,70 @@ impl<'tcx> GotocCtx<'tcx> {
-                                };
-                                // TODO: this assumes we have a slice. But it might be dynamic.
-                                slice_fat_ptr(
-                                    self.codegen_ty(res_ty),
-                                    // place.goto_expr.decay_to_ptr(),
-                                    data,
-                                    fat_ptr.member("len", &self.symbol_table),
-                                    &self.symbol_table,
-                                )
-                            }
-                            ty::Slice(_) | ty::Str | ty::Dynamic(..) => fat_ptr,
-                            _ => unreachable!(),
-                        }
-                    }
-                    _ => {
-                        // TODO: fully understand this case.
-                        // It seems that the thing to do here is just return the fat pointer we got from the place.
-                        fat_ptr
-                    }
-                }
-            }
-            ty::Bool | ty::Char | ty::Int(_) | ty::Uint(_) | ty::Float(_) => {
-                place.goto_expr.address_of()
-            }
-            ty::Adt(_, _)
-            | ty::Foreign(_)
-            | ty::Array(_, _)
-            | ty::RawPtr(_)
-            | ty::Ref(_, _, _)
-            | ty::FnDef(_, _)
-            | ty::FnPtr(_)
-            | ty::Closure(_, _)
-            | ty::Generator(_, _, _)
-            | ty::GeneratorWitness(_)
-            | ty::Never
-            | ty::Tuple(_)
-            | ty::Projection(_)
-            | ty::Opaque(_, _)
-            | ty::Param(_)
-            | ty::Bound(_, _)
-            | ty::Placeholder(_)
-            | ty::Infer(_)
-            | ty::Error(_) => {
-                // TODO: This matches what we had here before.
-                // But I'm not sure that its actually correct for all the match arms.
-                place.goto_expr.address_of()
-            }
-            ty::Slice(_) | ty::Str | ty::Dynamic(..) => unreachable!(),
+        debug!("codegen_rvalue_ref: place: {:?}", place);
+        debug!("codegen_rvalue_ref: place type: {:?}", place_mir_type);
+        debug!("codegen_rvalue_ref: place kind: {:?}", place_mir_type.kind());
+        debug!("codegen_rvalue_ref: projection: {:?}", projection);
+
+        assert!(
+            is_pointer(result_mir_type),
+            "Constructing a pointer of the type {:?} to the value of the place {:?}",
+            result_mir_type,
+            place
+        );
+        let result_goto_type = self.codegen_ty(result_mir_type);
+
+        // The goto expr for the value of this place
+        let place_goto_expr = projection.goto_expr;
+
+        // Construct a thin pointer to the value of this place
+        if self.use_thin_pointer(place_mir_type) {
+            return place_goto_expr.address_of();
+        }
+
+        // Construct a fat pointer to the value of this place
```

Comment:
It took me a second to realize that these are block comments for a bunch of code. Maybe distinguish that visually somehow?